java/ql/lib/semmle/code/java/controlflow/Guards.qll

```diff
@@ -108,8 +108,7 @@ class Guard extends ExprParent {
    */
   BasicBlock getBasicBlock() {
     result = this.(Expr).getBasicBlock() or
-    result = this.(SwitchCase).getSwitch().getExpr().getBasicBlock() or
-    result = this.(SwitchCase).getSwitchExpr().getExpr().getBasicBlock()
+    result = this.(SwitchCase).getSelectorExpr().getBasicBlock()
```

Comment:
This isn't quite what we want for pattern cases (and will also be wrong if we switch the cfg to sequential testing of const cases). This needs to return the basic block that has the outgoing true- and false-edges. I do realise that the existing code only approximated this for const cases if the selector expression wasn't a post-order cfg node, but for pattern cases we can do better.

Reply:
Also, the `hasBranchEdge` predicate below is wrong for pattern cases. But again, if we switch to a sequential switch cfg construction then it will need even further updates.

Reply:
I'll fix `getBasicBlock`, but can't expose this in a test at the moment because all uses of getBasicBlock either require the guard to be an equality test, or else are used to ask dominance questions (in either direction) where I can't see a way to expose the difference due to the top of the switch dominating the case, and the flow of alternate pattern cases only converging outside the switch block, because it is illegal to fall through either into or out of a pattern case.

Reply:
I've restricted `hasBranchEdge` to non-pattern cases, because pattern cases constitute ConditionBlocks now -- and if we make the switch-is-linear-series-of-tests change, const cases will be too.

Reply:
Looks like the only users where the guard could be a switch case are modulus and sign analysis;  I haven't immediately been able to cook a test to expose this logic.